### ISSUE TITLE
[import/no-unresolved] when using with typescript "baseUrl" and "paths" option

### ISSUE THREAD

**Issue by jjangga0214** [benmosher/eslint-plugin-import#1485]
# Repro

##### tree

```
<root>
├── src
│   ├── index.ts
│   └── foo.ts
├── tsconfig.json
├── .eslintrc.js
...
└── package.json
```

##### tsconfig.json

Other options are omiited for brevity.

```js
{
  "compilerOptions": {
    "baseUrl": "./src"
    "paths": {
      "#/*": ["./*"]
    }
  }
}
```
##### .eslintrc.js

Other options are omiited for brevity.

```js
module.exports = {
  parser: '@typescript-eslint/parser',
  root: true,
  env: {
    node: true,
    es6: true,
  },
  plugins: ['@typescript-eslint', ],
  extends: [
    'plugin:@typescript-eslint/recommended',
    'plugin:import/typescript',
  ],
}

```

##### src/index.ts

```ts
import { foo } from '#/foo'

console.log(`index.ts uses ${foo}`)
```

# Problem

**Expected Result**

It should respect `baseUrl` and `paths` option on tsconfig.json

**Actual Result**

```
error  Unable to resolve path to module '#/foo'  import/no-unresolved
```

**Additional Info**

<https://github.com/jjangga0214/ts-boilerplate/commit/af74c178a3c5ee84f78c520ba3c09ad73a9fc130/checks> is the actual result with an error message. (Expand it by clicking "Lint" on github action checks) 

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `2.2.0` |
| `@typescript-eslint/parser`        | `2.2.0` |
| `eslint-plugin-import`        | `2.18.2` |
| `TypeScript`                       | `3.6.3` |
| `ESLint`                           | `5.3.0` |
| `node`                             | `12.6.0` |
| `npm`                              | `6.10.2` |
| `yarn`                              | `1.17.3` |

**Comment by ljharb** [comment]
You'd have to use a typescript resolver for that, and configure it in your eslint settings.

See https://github.com/benmosher/eslint-plugin-import#resolvers and https://github.com/benmosher/eslint-plugin-import/wiki/Resolvers

**Comment by jjangga0214** [comment]
Thanks.

For other folks, this solves the issue.
 
```bash
npm install eslint-import-resolver-typescript
```

**package.json**

```js
{
  // other configuration are omitted for brevity
  settings: {
    "import/resolver": {
      typescript: {} // this loads <rootdir>/tsconfig.json to eslint
    },
  },
}
```
```

**Comment by alexkaduk** [comment]
@jjangga0214 
After `eslint-import-resolver-typescript` installation
`{
  // other configuration are omitted for brevity
  settings: {
    "import/resolver": {
      typescript: {} // this loads <rootdir>/tsconfig.json to eslint
    },
  },
}`
should be added to `.eslintrc.js`

**Comment by evanjmg** [comment]
This doesn't work, anyone else have a solution?

**Comment by Sergioamjr** [comment]
I needed to reopen my vscode but this worked very well, thanks
https://github.com/benmosher/eslint-plugin-import/issues/1485#issuecomment-571597574

**Comment by Sergioamjr** [comment]
> @jjangga0214
> After `eslint-import-resolver-typescript` installation
> 
> ```
> {
>   settings: {
>     "import/resolver": {
>       typescript: {} // this loads <rootdir>/tsconfig.json to eslint
>     },
>   },
> }
> ```
> 
> should be added to `.eslintrc.js`.
> Thanks.

I needed to reopen my vscode but this worked very well, thanks

**Comment by afuscella** [comment]
@jjangga0214  thanks for the suggestion. `eslint-import-resolver-typescript` worked just fine.

btw, I also added `alwaysTryTypes: true`, to force the resolve of @types first

cheers

**Comment by tpluscode** [comment]
I've already been using typescript resolver but it was necessary to add `"project": "packages/*/tsconfig.json"` to the resolver's config for it to work

**Comment by vortesnail** [comment]
> I've already been using typescript resolver but it was necessary to add `"project": "packages/*/tsconfig.json"` to the resolver's config for it to work

Yes!  

For monorepo projects, this is the required configuration. I was confused until I read your comment, thanks!

**Comment by jneander** [comment]
Is there a configuration, whether with `eslint-plugin-import` alone or in conjunction with other eslint extensions, which will ensure that the `.js` extension is present on imports of typescript files? Without another tool appending the extension as needed, it must be done manually to ensure that the `tsc`-outputted files are valid es modules. Without linter support, I know I will continue to miss extensions where they need to be.

**Comment by ljharb** [comment]
@jneander no, not that i know of. I'd suggest omitting the extension, and using babel instead of tsc as your transpiler (regardless), and then it'd be trivial to add a babel plugin that adds the extensions whenever needed.

**Comment by DalSoft** [comment]
This worked for me add this to tsconfig.json

```json
"settings": {
    "import/resolver": {
      "typescript": {}
    }
  }
```

**Comment by SalahAdDin** [comment]
> > @jjangga0214
> > After `eslint-import-resolver-typescript` installation
> > ```
> > {
> >   settings: {
> >     "import/resolver": {
> >       typescript: {} // this loads <rootdir>/tsconfig.json to eslint
> >     },
> >   },
> > }
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > should be added to `.eslintrc.js`.
> > Thanks.
> 
> I needed to reopen my vscode but this worked very well, thanks

Well, it does not work anymore, hours trying to make it work and I couldn't.

**Comment by peter-wd-1** [comment]
Dose not work

**Comment by kristijorgji** [comment]
Hi, it does not work for me using nextjs + typescript + eslint

My eslint file
```
module.exports = {
    root: true,
    parser: '@typescript-eslint/parser',
    parserOptions: {
        ecmaFeatures: { jsx: true },
    },
    extends: [
        'eslint:recommended',
        'plugin:@typescript-eslint/eslint-recommended',
        'plugin:@typescript-eslint/recommended',
        'plugin:react/recommended',
        'plugin:jsx-a11y/recommended',
        'plugin:import/errors',
        'plugin:import/warnings',
        'plugin:import/typescript',

        // Prettier plugin and recommended rules
        'plugin:prettier/recommended',
    ],
    rules: {
        // Include .prettierrc.js rules
        'prettier/prettier': ['error', {}, { usePrettierrc: true }],

        'no-empty': 'off',
        'react/prop-types': 'off',
        '@typescript-eslint/ban-ts-comment': 'off',
        '@typescript-eslint/no-non-null-assertion': 'off',
        'jsx-a11y/click-events-have-key-events': 'off',
        'jsx-a11y/no-static-element-interactions': 'off',
        'jsx-a11y/interactive-supports-focus': 'off',
        'jsx-a11y/no-noninteractive-element-interactions': 'off',
        'jsx-a11y/no-onchange': 'off',
    },
    settings: {
        'import/extensions': ['.js', '.jsx', '.tsx', '.ts'],
        'import/resolver': {
            typescript: {},
        },
    },
};
```

In tsconfig.json i use `paths` 
```
"paths": {
      "@/core/*": ["src/core/*"],
}
```
and get errors from eslint like

```
  18:23  error  Unable to resolve path to module '@/core/useRouter'                  import/no-unresolved

```

Compilation build all works great, just linting fails eslint

**Comment by kristijorgji** [comment]
Hi, it does not work for me using nextjs + typescript + eslint

My eslint file
```
module.exports = {
    root: true,
    parser: '@typescript-eslint/parser',
    parserOptions: {
        ecmaFeatures: { jsx: true },
    },
    extends: [
        'eslint:recommended',
        'plugin:@typescript-eslint/eslint-recommended',
        'plugin:@typescript-eslint/recommended',
        'plugin:react/recommended',
        'plugin:jsx-a11y/recommended',
        'plugin:import/errors',
        'plugin:import/warnings',
        'plugin:import/typescript',

        // Prettier plugin and recommended rules
        'plugin:prettier/recommended',
    ],
    rules: {
        // Include .prettierrc.js rules
        'prettier/prettier': ['error', {}, { usePrettierrc: true }],

        'no-empty': 'off',
        'react/prop-types': 'off',
        '@typescript-eslint/ban-ts-comment': 'off',
        '@typescript-eslint/no-non-null-assertion': 'off',
        'jsx-a11y/click-events-have-key-events': 'off',
        'jsx-a11y/no-static-element-interactions': 'off',
        'jsx-a11y/interactive-supports-focus': 'off',
        'jsx-a11y/no-noninteractive-element-interactions': 'off',
        'jsx-a11y/no-onchange': 'off',
    },
    settings: {
        'import/extensions': ['.js', '.jsx', '.tsx', '.ts'],
        'import/resolver': {
            typescript: {},
        },
    },
};
```

In tsconfig.json i use `paths` 
```
"paths": {
      "@/core/*": ["src/core/*"],
}
```
and get errors from eslint like

```
  18:23  error  Unable to resolve path to module '@/core/useRouter'                  import/no-unresolved

```

Compilation build all works great, just linting fails eslint

**Comment by ljharb** [comment]
Sounds like a new issue would be appropriate; commenting on a closed one isn’t likely to help anything.

**Comment by kristijorgji** [comment]
@ljharb good idea, I created this one you guys can share your samples there 
https://github.com/import-js/eslint-plugin-import/issues/2405

**Comment by meiqiududu** [comment]
const path = require('path');

`settings: {
    'import/resolver': {
      typescript: {
        // directory: './tsconfig.json', //该写法无效
        project: path.join(__dirname, './tsconfig.json'), // 需安装npm install eslint-import-resolver-typescript 插件读取tsconfig配置的路径，关键配置
        alwaysTryTypes: true
      },
      node: {
        extensions: ['.js', '.jsx', '.ts', '.tsx']
      }
    }
  }`

**Comment by mhfaust** [comment]
By the way, by now it seems if you've got a TS app in VSCode with your `paths` configured correctly in tsconfig.json compilerOptions, you don't really need this rule and you can just disable it. The integrated compiler always catches typos in import paths.

Therefore your best option is just:
```
rules: [
   "import/no-unresolved": "off",
]
```

It took me a while to figure that out, and I kicked myself when I did.

**Comment by king-koda** [comment]
> By the way, by now it seems if you've got a TS app in VSCode with your `paths` configured correctly in tsconfig.json compilerOptions, you don't really need this rule and you can just disable it. The integrated compiler always catches typos in import paths.
> 
> Therefore your best option is just:
> 
> ```
> rules: [
>    "import/no-unresolved": "off",
> ]
> ```
> 
> It took me a while to figure that out, and I kicked myself when I did.

thank you bro, I did wonder why I'm spending 5+ hours trying to fix this rubbish when if the pathing was broken TS would of told me, I should of not been so stubborn, set and forget, how miserable HAHA

**Comment by SalahAdDin** [comment]
@jjangga0214  if this issue is still happening, it should not be solved.

**Comment by jjangga0214** [comment]
@SalahAdDin What do you mean? Sorry, but I don't understand your message.

**Comment by SalahAdDin** [comment]
> > By the way, by now it seems if you've got a TS app in VSCode with your `paths` configured correctly in tsconfig.json compilerOptions, you don't really need this rule and you can just disable it. The integrated compiler always catches typos in import paths.
> > Therefore your best option is just:
> > ```
> > rules: [
> >    "import/no-unresolved": "off",
> > ]
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > It took me a while to figure that out, and I kicked myself when I did.
> 
> thank you bro, I did wonder why I'm spending 5+ hours trying to fix this rubbish when if the pathing was broken TS would of told me, I should of not been so stubborn, set and forget, how miserable HAHA

One hour ago, a new answer, so, the issue still happends.

**Comment by powderham** [comment]
> > By the way, by now it seems if you've got a TS app in VSCode with your `paths` configured correctly in tsconfig.json compilerOptions, you don't really need this rule and you can just disable it. The integrated compiler always catches typos in import paths.
> > Therefore your best option is just:
> > ```
> > rules: [
> >    "import/no-unresolved": "off",
> > ]
> > ```
> > 
> > 
> >     
> >   
> > It took me a while to figure that out, and I kicked myself when I did.
> 
> thank you bro, I did wonder why I'm spending 5+ hours trying to fix this rubbish when if the pathing was broken TS would of told me, I should of not been so stubborn, set and forget, how miserable HAHA

If you're going to do this, best to turn it off only on ts(x) files

```
overrides: [
    {
      files: ["*.ts", "*.tsx"],
      rules: {
        "import/no-unresolved": "off",
      },
    },
  ],
```
